### ISSUE TITLE
Serialization and deserialization issue of sub-types used with `JsonTypeInfo.Id.DEDUCTION` where sub-types are Object and Array

### ISSUE THREAD

**Issue by Gems** [FasterXML/jackson-databind#4772]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

Having the following class structure and annotations the deserialization and serialization doesn't work (but expected):

```
    @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION)
    @JsonSubTypes({@JsonSubTypes.Type(DataObject.class), @JsonSubTypes.Type(DataArray.class)})
    interface Data {
    }

    static class DataItem {
        String id;
    }

    static class DataObject extends DataItem implements Data {
       @JsonCreator
        DataObject(@JsonProperty("id") String id) {
            super(id);
        }
    }

    static class DataArray extends ArrayList<DataItem> implements Data {
        @JsonCreator
        DataArray(Collection<DataItem> items) {
            super(new ArrayList<>(items));
        }
    }

    static class Container {
        Data data;
    }
```

### Version Information

2.18.0
and, I believe, 2.19

### Reproduction

Here is a test case that I made passing: https://github.com/Gems/jackson-databind/blob/contrib/2.18-1/src/test/java/com/fasterxml/jackson/databind/jsontype/TestPolymorphicDeductionObjectVsArray.java

To make Deserialization work, I had to change `@JsonTypeInfo` annotation to `@JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION, defaultImpl = DataArray.class)` 

To make Serialization work, I had to fix `JsonGenerator::writeTypePrefix` method (comes from jackson-core) to ignore making `WRAPPER_ARRAY` in case `typeIdDef.id == null`.

### Expected behavior

I'd expect that with the mentioned definition the deserialization and serialization goes smooth without tweaks that I made, but not sure if this expectation is aligned with the vision for the library design.

Thanks in advance for consideration and comments.

**Comment by cowtowncoder** [comment]
Yeah, polymorphic deduction can not work on non-POJO types such as `Collection`s, `Map`s, arrays, primitives/scalars as it is based on concept of Properties.

So this is not supported use case unfortunately.

Would be good to have improvements in Javadocs; will see what is said about `DEDUCTION`.

**Comment by Gems** [comment]
Thanks for your feedback @cowtowncoder 

I wonder if we make at least "fix" `JsonGenerator::writeTypePrefix`. I believe producing `"null"` as `TypeDef.id` doesn't suit to any use case. What would you say?

**Comment by cowtowncoder** [comment]
@Gems I am not sure what that comment refers to in the context of this issue.

**Comment by cowtowncoder** [comment]
@Gems If you want to do PR against 2.19 (this is probably too risky for patch), I can have a look. Implementation may be incorrect for the case you describe, it's quite hairy all in all and not well tested for edge cases (like no Type Id).

**Comment by Gems** [comment]
@cowtowncoder, gotcha, I'll do the patch for Jackson-core 2.19, will run the tests, and see. If it's all green then I'll open a PR.

Thanks!

**Comment by cowtowncoder** [comment]
@Gems Although the fix in jackson-core will only go in 2.19, it should be an improvement and possible fix other existing issues. But more importantly does it (eventually) fix your original issue? :)

**Comment by Gems** [comment]
Yes, it does. Here is a PR with added tests relevant to the issue 👍🏻
